src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -25,20 +27,413 @@ using neural_net::model_backend;
 using neural_net::model_spec;
 using neural_net::shared_float_array;
 
+struct result {
+  shared_float_array loss_info;
+  shared_float_array accuracy_info;
+  data_iterator::batch data_info;
+};
+
 }  // namespace
 
-std::unique_ptr<model_spec> drawing_classifier::init_model() const {
+std::unique_ptr<model_spec> drawing_classifier::init_model(
+    bool use_random_init) const {
   std::unique_ptr<model_spec> result(new model_spec);
   return result;
 }
 
+void drawing_classifier::init_options(
+    const std::map<std::string, flexible_type>& opts) {
+  // Define options.
+  options.create_integer_option(
+      "batch_size",
+      "Number of training examples used per training step",
+      256,
+      1,
+      std::numeric_limits<int>::max());
+  options.create_integer_option(
+      "max_iterations",
+      "Maximum number of iterations/epochs made over the data during the"
+      " training phase",
+      500,
+      1,
+      std::numeric_limits<int>::max());
+  
+  // Validate user-provided options.
+  options.set_options(opts);
+
+  // Write model fields.
+  add_or_update_state(flexmap_to_varmap(options.current_option_values()));
+}
+
+std::tuple<gl_sframe, gl_sframe> drawing_classifier::init_data(
+      gl_sframe data, variant_type validation_data) const {
+  gl_sframe train_data;
+  gl_sframe val_data;
+  if (variant_is<gl_sframe>(validation_data)) {
+    train_data = data;
+    val_data = variant_get_value<gl_sframe>(validation_data);
+    /* TODO: Do we want to log_and_throw when the validation data is empty?
+     *       Having it will trigger a unit test failure!
+     */
+  }
+  else if ((variant_is<flex_string>(validation_data)) 
+      && (variant_get_value<flex_string>(validation_data)=="auto")) {
+    size_t num_rows = data.size();
+    if (num_rows >= 100) {
+      std::tie(train_data, val_data) = data.random_split(.95);
+    } else {
+      train_data = data;
+      std::cout << "The dataset has less than the minimum of 100 rows required for train-validation split. "
+                       "Continuing without validation set.\n";
+    }
```

Comment:
Could we just use this implementation instead: https://github.com/apple/turicreate/blob/d06876920a172e41aa95341f0d2392d26e01d85b/src/toolkits/supervised_learning/automatic_model_creation.hpp#L28
?